src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -461,11 +459,11 @@ private static string GetPreferredUsernameFromIdToken(bool isAdfsAuthority, IdTo
             string partitionKey = CacheKeyFactory.GetKeyFromRequest(requestParams);
             Debug.Assert(partitionKey != null || !requestParams.IsConfidentialClient, "On confidential client, cache must be partitioned.");
 
-            IReadOnlyList<MsalAccessTokenCacheItem> tokenCacheItems = GetAllAccessTokensWithNoLocks(true, partitionKey);
+            var accessTokens = GetAllAccessTokensWithNoLocks(true, partitionKey);
```

Comment:
I'm thinking that this filter here is wrong. We get all tokens and filter by client_id, but it's highly unlikely that there will be more than 1 client id in the whole cache.

So we are not filtering by the most restrictive filter first, which for S2S is the tenant id. `GetAllAccessTokensWithNoLocks` should not even have this flag to filter by client id, as it promotes misuse.

Thoughts?

Reply:
So your suggestion is to still keep client ID filter, and just move it down the filtering list?

For client credential calls, the partition includes client ID, so this filter is unnecessary. For user related calls, the partition is per user, which could have tokens for different client IDs (whether the cache is static or serialized), right? But yes, the filter is the least restrictive, and probably least likely to happen.

Also `WithNoLocks` in the method name is unnecessary, since we never lock when we use accessors anymore.

Reply:
@bgavrilMS Changes made in 292de9f2312ce83e8afe7a264054549322d8a0c6.

Reply:
It's a good point about the partition key.